File: manifests/emqx-iot/simulator/api/db/bootstrap.js
import { readdir, readFile } from "node:fs/promises";

const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

export function createBootstrapService({ db, config }) {
  async function init() {
    await db.query("CREATE EXTENSION IF NOT EXISTS timescaledb");

    const scriptNames = (await readdir(config.bootstrapSqlDir))
      .filter((name) => name.endsWith(".sql"))
      .sort((a, b) => a.localeCompare(b, undefined, { numeric: true }));

    for (const scriptName of scriptNames) {
      const script = await readFile(`${config.bootstrapSqlDir}/${scriptName}`, "utf8");
      if (!script.trim()) continue;

      if (scriptName === "04-gold-continuous-aggregates.sql") {
        const statements = script
          .split(/;\s*\n/g)
          .map((part) => part.trim())
          .filter(Boolean);

        for (const statement of statements) {
          await db.query(`${statement};`);
        }
        continue;
      }

      await db.query(script);
    }
  }

  async function initWithRetry() {
    for (let attempt = 1; attempt <= config.bootstrapMaxAttempts; attempt += 1) {
      try {
        await init();
        return;
      } catch (error) {
        if (attempt === config.bootstrapMaxAttempts) throw error;
        console.error(`database bootstrap attempt ${attempt} failed: ${error.message}`);
        await sleep(config.bootstrapRetryDelayMs);
      }
    }
  }

  return { init, initWithRetry };
}

File: manifests/emqx-iot/simulator/api/app/bootstrap.py
import asyncio
from pathlib import Path

from app.infrastructure.persistence.db import db
from app.settings import settings


async def init_db() -> None:
    await db.execute("CREATE EXTENSION IF NOT EXISTS timescaledb")

    sql_dir = Path(settings.db_bootstrap_sql_dir)
    for path in sorted(sql_dir.glob("*.sql")):
        script = path.read_text(encoding="utf-8").strip()
        if not script:
            continue

        if path.name == "04-gold-continuous-aggregates.sql":
            statements = [s.strip() for s in script.split(";\n") if s.strip()]
            for statement in statements:
                await db.execute(f"{statement};")
        else:
            await db.execute(script)


async def init_db_with_retry() -> None:
    for attempt in range(1, settings.db_bootstrap_max_attempts + 1):
        try:
            await init_db()
            return
        except Exception:
            if attempt == settings.db_bootstrap_max_attempts:
                raise
            await asyncio.sleep(settings.db_bootstrap_retry_delay_ms / 1000)
